python/paddle/fluid/dygraph/dygraph_to_static/loop_transformer.py

```diff
@@ -0,0 +1,251 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import copy
+import gast
+
+from collections import defaultdict
+from paddle.fluid import unique_name
+from paddle.fluid.dygraph.dygraph_to_static.ast_utils import create_funcDef_node
+from paddle.fluid.dygraph.dygraph_to_static.ast_utils import generate_name_node
+from paddle.fluid.dygraph.dygraph_to_static.static_analysis import AstNodeWrapper
+from paddle.fluid.dygraph.dygraph_to_static.variable_trans_func import create_static_variable_gast_node
+from paddle.fluid.dygraph.dygraph_to_static.variable_trans_func import to_static_variable_gast_node
+
+__all__ = ['LoopTransformer', 'NameVisitor']
+
+WHILE_CONDITION_PREFIX = 'while_condition'
+WHILE_BODY_PREFIX = 'while_body'
+
+
+def create_while_node(condition_name, body_name, loop_var_names):
+    while_args = []
+    while_args.append(
+        gast.Name(
+            id=condition_name,
+            ctx=gast.Param(),
+            annotation=None,
+            type_comment=None))
+    while_args.append(
+        gast.Name(
+            id=body_name, ctx=gast.Param(), annotation=None, type_comment=None))
+    assign_targets = [
+        gast.Name(
+            id=var_name, ctx=gast.Param(), annotation=None, type_comment=None)
+        for var_name in loop_var_names
+    ]
+    while_args.append(gast.List(elts=assign_targets, ctx=gast.Param()))
+
+    while_func_id = gast.parse('fluid.layers.while_loop').body[0].value
+    while_node = gast.Call(func=while_func_id, args=while_args, keywords=[])
+    assign_node = gast.Assign(
+        targets=[gast.Tuple(
+            elts=assign_targets, ctx=gast.Store())],
+        value=while_node)
+    return assign_node
+
+
+class NameVisitor(gast.NodeVisitor):
+    '''
+    Analysis name liveness for loop transformer
+    '''
+
+    def __init__(self, root_node):
+        # Set of gast.Name
+        self.current_seen_vars = set()
+        # List of gast.While/gast.For nodes
+        self.current_loop = []
+
+        # Mapping from gast.While/gast.For to string name of vars
+        self.before_loop_vars = defaultdict(set)
+        self.in_loop_vars = defaultdict(set)
+
+        self.visit(root_node)
+
+    def is_control_flow_loop(self, node):
+        # TODO: make a better condition
+        return True
+
+    def get_loop_var_names(self, node):
+        assert isinstance(node, gast.While) or isinstance(
+            while_node, gast.For), "Input node is not gast loop node"
```

Comment:
I didn't find where the `while_node` is defined? Is it means `node`?

If does, `assert isinstance(node, (gast.While, gst.For))` will be better. 😁

Reply:
Oh, that's a bug and the test didn't go to the For case so it wasn't catch

Reply:
Oh, that's a bug and the test didn't go to the For case so it wasn't catch.

I will modify in next PR